File: kotlin/src/main/AoC2017/day19.kt
package main.AoC2017

import java.io.File


fun main(args: Array<String>){
    
    
    val grid = File("inputs/2017/day19.in").readLines().map{it.toCharArray()}.toList()
    val SOUTH = 0
    val EAST = 1
    val NORTH = 2
    val WEST = 3
    var direction = SOUTH
    var row = 0
    var col = grid[0].indexOf('|')
    var result = ""
    var steps = 0
    while(grid[row][col] != ' ') {
        steps++
        if(grid[row][col].isLetter()){
            result += grid[row][col]
        }else if(grid[row][col] == '+'){
            direction = if(direction % 2 == 0){
                if(grid[row][col + 1] != ' '){
                    EAST
                }else{
                    WEST
                }
            }else{
                if(grid[row + 1][col] != ' '){
                    SOUTH
                }else{
                    NORTH
                }
            }
        }
        when(direction){
            SOUTH -> row++
            NORTH -> row--
            EAST  -> col++
            WEST  -> col--
        }

    }
    
    println("Part 1: $result")
    println("Part 2: $steps")
    
}

File: python/Day19/day19.py
import re, hashlib, itertools
from collections import *

inpt = open('day19.in').read()
grid = [[char for char in line] for line in inpt.splitlines()]
row, column = 0, grid[0].index('|')
direction = 'south'
moves = {'south': (1, 0), 'east':(0,1), 'north':(-1, 0), 'west':(0, -1)}
encountered = ''
steps = 0

while grid[row][column] != ' ':
    steps += 1
    if grid[row][column].isalpha():
        encountered += grid[row][column]
    elif grid[row][column] == '+':
        if direction in ['south', 'north']:
            if grid[row][column + 1] != ' ':
                direction = 'east'
            else:
                direction = 'west'
        else:
            if grid[row + 1][column] != ' ':
                direction = 'south'
            else:
                direction = 'north'

    dr, dc = moves[direction]
    row += dr
    column += dc


print('Part 1: {}'.format(encountered))
print('Part 2: {}'.format(steps))






